src/DIRAC/FrameworkSystem/Agent/ComponentSupervisionAgent.py

```diff
@@ -0,0 +1,549 @@
+
+# imports
+from collections import defaultdict
+from datetime import datetime
+from functools import partial
+
+import os
+import socket
+import psutil
+
+# from DIRAC
+from DIRAC import S_OK, S_ERROR, gConfig
+from DIRAC.Core.Base.AgentModule import AgentModule
+from DIRAC.Core.Base.Client import Client
+from DIRAC.Core.Utilities.PrettyPrint import printTable
+from DIRAC.FrameworkSystem.Client.SystemAdministratorClient import SystemAdministratorClient
+from DIRAC.ConfigurationSystem.Client.CSAPI import CSAPI
+from DIRAC.ConfigurationSystem.Client.Helpers.Path import cfgPath
+from DIRAC.FrameworkSystem.Client.NotificationClient import NotificationClient
+from DIRAC.WorkloadManagementSystem.Client.JobMonitoringClient import JobMonitoringClient
+
+__RCSID__ = "$Id$"
+AGENT_NAME = 'Framework/ComponentSupervisionAgent'
+
+# Define units
+HOUR = 3600
+MINUTES = 60
+SECONDS = 1
+
+# Define constant
+NO_CHECKING_JOBS = 'NO_CHECKING_JOBS'
+CHECKING_JOBS = 'CHECKING_JOBS'
+NO_RESTART = 'NO_RESTART'
+
+
+class ComponentSupervisionAgent(AgentModule):
+  """ComponentSupervisionAgent class."""
+
+  def __init__(self, *args, **kwargs):
+    """Initialize the agent, clients, default values."""
+    AgentModule.__init__(self, *args, **kwargs)
+    self.name = 'ComponentSupervisionAgent'
+    self.setup = 'Production'
+    self.enabled = False
+    self.restartAgents = False
+    self.restartExecutors = False
+    self.restartServices = False
+    self.controlComponents = False
+    self.commitURLs = False
+    self.doNotRestartInstancePattern = ['RequestExecutingAgent']
+    self.diracLocation = "/opt/dirac/pro"
+
+    self.sysAdminClient = SystemAdministratorClient(socket.gethostname())
+    self.jobMonClient = JobMonitoringClient()
+    self.nClient = NotificationClient()
+    self.csAPI = None
+    self.agents = dict()
+    self.executors = dict()
+    self.services = dict()
+    self.errors = list()
+    self.accounting = defaultdict(dict)
+
+    self.addressTo = []
+    self.addressFrom = ''
+    self.emailSubject = 'ComponentSupervisionAgent on %s' % socket.gethostname()
+
+  def logError(self, errStr, varMsg=''):
+    """Append errors to a list, which is sent in email notification."""
+    self.log.error(errStr, varMsg)
+    self.errors.append(errStr + ' ' + varMsg)
+
+  def beginExecution(self):
+    """Reload the configurations before every cycle."""
+    self.setup = self.am_getOption('Setup', self.setup)
+    self.enabled = self.am_getOption('EnableFlag', self.enabled)
+    self.restartAgents = self.am_getOption('RestartAgents', self.restartAgents)
+    self.restartExecutors = self.am_getOption('RestartExecutors', self.restartExecutors)
+    self.restartServices = self.am_getOption('RestartServices', self.restartServices)
+    self.diracLocation = os.environ.get('DIRAC', self.diracLocation)
+    self.addressTo = self.am_getOption('MailTo', self.addressTo)
+    self.addressFrom = self.am_getOption('MailFrom', self.addressFrom)
```

Comment:
If these are not defined, maybe you take them from the Operations/Shifter location?

Reply:
which shifter?